File: src/main/java/semicontinuity/idea/code/analyzer/graph/viewModel/ide/BaseIdeGraphViewFactory.java
package semicontinuity.idea.code.analyzer.graph.viewModel.ide;

import java.awt.Color;
import java.util.List;

import javax.swing.BorderFactory;
import javax.swing.Box;
import javax.swing.JComponent;

import semicontinuity.idea.code.analyzer.graph.viewModel.Factory;

public abstract class BaseIdeGraphViewFactory<VERTEX_PAYLOAD> implements Factory<
        VERTEX_PAYLOAD,
        JComponent,
        JComponent,
        JComponent,
        JComponent,
        JComponent
        > {
    static boolean showDebugBorders = false;

    @Override
    public JComponent newIndependentComponents(List<? extends JComponent> components) {
        var box = Box.createVerticalBox();
        if (showDebugBorders) {
            box.setBorder(BorderFactory.createLineBorder(Color.BLUE));
        } else {
            box.setBorder(BorderFactory.createEmptyBorder(1, 1, 1, 1));
        }

        for (JComponent component : components) {
            box.add(component);
        }
        return box;
    }

    @Override
    public JComponent newSplit(List<JComponent> items, JComponent subLayer) {
        var box = Box.createHorizontalBox();
        if (showDebugBorders) {
            box.setBorder(BorderFactory.createLineBorder(Color.YELLOW));
        } else {
            box.setBorder(BorderFactory.createEmptyBorder(1, 1, 1, 1));
        }

        box.add(itemsBox(items));
        box.add(subLayer);
        return box;
    }

    @Override
    public JComponent newLayer(JComponent directDeps, JComponent sharedDeps) {
        var box = Box.createVerticalBox();
        if (showDebugBorders) {
            box.setBorder(BorderFactory.createLineBorder(Color.DARK_GRAY));
        } else {
            box.setBorder(BorderFactory.createEmptyBorder(1, 1, 1, 1));
        }

        if (directDeps != null) {
            box.add(directDepsBox(directDeps));
        }
        if (sharedDeps != null) {
            box.add(sharedDepsBox(sharedDeps));
        }
        return box;
    }

    JComponent itemsBox(List<JComponent> items) {
        var box = Box.createVerticalBox();
        if (showDebugBorders) {
            box.setBorder(BorderFactory.createLineBorder(Color.PINK));
        } else {
            box.setBorder(BorderFactory.createEmptyBorder(1, 1, 1, 1));
        }

        for (JComponent component : items) {
            box.add(component);
        }
        return box;
    }

    JComponent directDepsBox(JComponent contents) {
        var box = Box.createVerticalBox();
        if (showDebugBorders) {
            box.setBorder(BorderFactory.createLineBorder(Color.GREEN));
        } else {
            box.setBorder(BorderFactory.createEmptyBorder(1, 1, 1, 1));
        }

        box.add(contents);
        return box;
    }

    JComponent sharedDepsBox(JComponent contents) {
        var box = Box.createVerticalBox();
        if (showDebugBorders) {
            box.setBorder(BorderFactory.createLineBorder(Color.RED));
        } else {
            box.setBorder(BorderFactory.createLoweredBevelBorder());
        }

        box.add(contents);
        return box;
    }
}

File: src/main/java/semicontinuity/idea/code/analyzer/graph/viewModel/ide/BaseIdeGraphViewFactory.kt
package semicontinuity.idea.code.analyzer.graph.viewModel.ide

import com.intellij.ui.JBColor
import semicontinuity.idea.code.analyzer.graph.viewModel.Factory
import java.awt.Color
import javax.swing.BorderFactory
import javax.swing.Box
import javax.swing.JComponent

abstract class BaseIdeGraphViewFactory<VERTEX_PAYLOAD> :
    Factory<VERTEX_PAYLOAD, JComponent, JComponent, JComponent, JComponent, JComponent, JComponent> {

    override fun newIndependentComponents(components: List<JComponent?>): JComponent {
        val box = Box.createVerticalBox()
        if (showDebugBorders) {
            box.border = BorderFactory.createLineBorder(Color.BLUE)
        } else {
            box.border = BorderFactory.createEmptyBorder(1, 1, 1, 1)
        }

        box.add(Box.createVerticalGlue())
        for (component in components) {
            box.add(component)
        }
        box.add(Box.createVerticalGlue())
        return box
    }

    override fun newSplit(items: List<JComponent>, subLayer: JComponent) =
        Box.createHorizontalBox().apply {
            add(itemsBox(items))
            add(Box.createHorizontalStrut(4))
            add(subLayer)
            add(Box.createHorizontalGlue())

            border = if (showDebugBorders) {
                BorderFactory.createLineBorder(Color.YELLOW)
            } else {
                BorderFactory.createEmptyBorder(1, 1, 1, 1)
            }
        }

    override fun newSplit(left: JComponent, right: JComponent) =
        Box.createHorizontalBox().apply {
            // setBorder(border(JBColor.YELLOW))
            setBorder(BorderFactory.createLineBorder(JBColor.LIGHT_GRAY))
            add(left)
            add(right)
        }

    override fun newLayer(directDeps: JComponent?, sharedDeps: JComponent?): JComponent {
        val box = Box.createVerticalBox()
        if (showDebugBorders) {
            box.border = BorderFactory.createLineBorder(Color.DARK_GRAY)
        } else {
            box.border = BorderFactory.createEmptyBorder(1, 1, 1, 1)
        }

        box.add(Box.createVerticalGlue())

        if (directDeps != null) {
            box.add(directDepsBox(directDeps))
        }
        if (sharedDeps != null) {
            box.add(sharedDepsBox(sharedDeps))
        }

        box.add(Box.createVerticalGlue())

        return box
    }

    fun itemsBox(items: List<JComponent>): JComponent {
        val box = Box.createVerticalBox()
        if (showDebugBorders) {
            box.border = BorderFactory.createLineBorder(Color.PINK)
        } else {
            box.border = BorderFactory.createEmptyBorder(1, 1, 1, 1)
        }

        box.add(Box.createVerticalGlue())
        for (component in items) {
            box.add(component)
        }
        box.add(Box.createVerticalGlue())
        return box
    }

    fun directDepsBox(contents: JComponent?): JComponent {
        val box = Box.createVerticalBox()

        if (showDebugBorders) {
            box.border = BorderFactory.createLineBorder(Color.GREEN)
        } else {
            box.border = BorderFactory.createEmptyBorder(1, 1, 1, 1)
        }

        box.add(Box.createVerticalGlue())
        box.add(contents)
        box.add(Box.createVerticalGlue())
        return box
    }

    fun sharedDepsBox(contents: JComponent?): JComponent {
        val box = Box.createVerticalBox()
        if (showDebugBorders) {
            box.border = BorderFactory.createLineBorder(Color.RED)
        } else {
            box.border = BorderFactory.createLoweredBevelBorder()
        }

        box.add(Box.createVerticalGlue())
        box.add(contents)
        box.add(Box.createVerticalGlue())
        return box
    }

    companion object {
        var showDebugBorders: Boolean = false
    }
}
